### PR TITLE
Revamp oidc endpoints retrieval

### PR DESCRIPTION
**Pull request by yathindrakodithuwakku**
## Approach
fixes https://github.com/wso2-enterprise/asgardeo-product/issues/9645

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: lib/src/client.ts [hunk 1]
```diff
@@ -17,6 +17,7 @@
  */
 
 import { OIDC_SCOPE, OP_CONFIG_INITIATED, ResponseMode, SIGN_OUT_SUCCESS_PARAM, STATE } from "./constants";
+import { SERVER_ENVIRONMENTS } from "./constants/server-environments";
```
Comment (thivi) [inline lib/src/client.ts:20]: We can merge this together.

File: lib/src/helpers/authentication-helper.ts [hunk 2]
```diff
@@ -33,8 +34,10 @@ import {
     SERVICE_RESOURCES,
     TOKEN_ENDPOINT,
     TOKEN_TAG,
+    USERINFO_ENDPOINT,
     USERNAME_TAG
 } from "../constants";
+import { SERVER_ENVIRONMENTS } from "../constants/server-environments";
```
Comment (thivi) [inline lib/src/helpers/authentication-helper.ts:40]: Shall we import this from the `"../constants"`?

File: lib/src/helpers/authentication-helper.ts [hunk 3]
```diff
@@ -62,56 +65,104 @@ export class AuthenticationHelper<T> {
...
-                        : "";
-                });
+        const requiredEndpoints = [AUTHORIZATION_ENDPOINT, END_SESSION_ENDPOINT, JWKS_ENDPOINT, 
+            OIDC_SESSION_IFRAME_ENDPOINT, REVOKE_TOKEN_ENDPOINT, TOKEN_ENDPOINT, ISSUER, USERINFO_ENDPOINT];
+
+        const isRequiredEndpointsContains = configData.endpoints ? 
+                Object.keys(configData?.endpoints).every(endpointName => {
+            const snakeCasedName = endpointName.replace(/[A-Z]/g, (letter) => `_${ letter.toLowerCase() }`);
+            return requiredEndpoints.includes(snakeCasedName);
+        }) : false;
```
Comment (thivi) [inline lib/src/helpers/authentication-helper.ts:94]: Shall we format this section?

File: lib/src/models/client-config.ts [hunk 4]
```diff
@@ -18,8 +18,9 @@
 
 import { OIDCEndpoints } from ".";
```
Comment (thivi) [inline lib/src/models/client-config.ts:19]: Possible circular dependency. We need to use a more specific path.

File: lib/src/models/client-config.ts [hunk 5]
```diff
@@ -49,7 +46,24 @@ export interface StrictAuthClientConfig {
...
+    endpoints?: OIDCEndpoints;
+}
+
+export interface OrganizationAuthClientConfig extends DefaultAuthClientConfig {
+    organization: string;
+    endpoints?: OIDCEndpoints;
 }
 
+export interface ExplicitAuthClientConfig extends DefaultAuthClientConfig {
+    endpoints: OIDCEndpoints;
```
Comment (thivi) [inline lib/src/models/client-config.ts:63]: Shall we make the required endpoints mandatory as well?

File: lib/src/helpers/authentication-helper.ts [hunk 6]
```diff
@@ -62,56 +65,106 @@ export class AuthenticationHelper<T> {
...
+        
         const defaultEndpoints = {
-            [ AUTHORIZATION_ENDPOINT ]: configData.serverOrigin + SERVICE_RESOURCES.authorizationEndpoint,
-            [ END_SESSION_ENDPOINT ]: configData.serverOrigin + SERVICE_RESOURCES.endSessionEndpoint,
-            [ JWKS_ENDPOINT ]: configData.serverOrigin + SERVICE_RESOURCES.jwksUri,
-            [ OIDC_SESSION_IFRAME_ENDPOINT ]: configData.serverOrigin + SERVICE_RESOURCES.checkSessionIframe,
-            [ REVOKE_TOKEN_ENDPOINT ]: configData.serverOrigin + SERVICE_RESOURCES.revocationEndpoint,
-            [ TOKEN_ENDPOINT ]: configData.serverOrigin + SERVICE_RESOURCES.tokenEndpoint
+            [ AUTHORIZATION_ENDPOINT ]: this.constructServerEndpoint(environment, 
+                organization ,SERVICE_RESOURCES.authorizationEndpoint),
```
Comment (thivi) [inline lib/src/helpers/authentication-helper.ts:142]:
```suggestion
                organization, SERVICE_RESOURCES.authorizationEndpoint),
```
```suggestion
                organization ,SERVICE_RESOURCES.authorizationEndpoint),
```

File: lib/src/models/client-config.ts [hunk 7]
```diff
@@ -49,7 +46,35 @@ export interface StrictAuthClientConfig {
...
+        authorizationEndpoint: string;
+        tokenEndpoint: string;
+        userinfoEndpoint: string;
+        jwksUri: string;
+        registrationEndpoint: string;
+        revocationEndpoint: string;
+        introspectionEndpoint: string;
+        checkSessionIframe: string;
+        endSessionEndpoint: string;
+        issuer: string;
```
Comment (thivi) [inline lib/src/models/client-config.ts:73]: Can't we use `OIDCEndpoints`?

Comment (thivi) [inline lib/src/models/client-config.ts:73]: Can't we use `OIDCEndpoints`?

File: lib/src/constants/server-environments.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
+/**
+ * Copyright (c) 2022, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
```
Comment (UdeshAthukorala) [inline lib/src/constants/server-environments.ts:9]: Can we remove this additional space here?

File: lib/src/models/client-config.ts [hunk 9]
```diff
@@ -49,7 +46,24 @@ export interface StrictAuthClientConfig {
      * Note that if this is set to false, then the needed endpoints should be provided through
      * the `endpoints` property.
      */
-    sendWellKnownEndpointRequest?: boolean;
+     environment?: SERVER_ENVIRONMENTS;
```
Comment (UdeshAthukorala) [inline lib/src/models/client-config.ts:49]: Can we remove this additional space here?

File: lib/src/constants/server-environments.ts (hunk 8 above)
Comment (yathindrakodithuwakku) [reply lib/src/constants/server-environments.ts:9]: The `lib/src/constants/server-environments.ts` file has been removed now, as we are not using environment-specific logic.

File: lib/src/models/client-config.ts (hunk 9 above)
Comment (yathindrakodithuwakku) [reply lib/src/models/client-config.ts:49]: The specified `environment` field is removed as the approach has been changed.

**Pull request merged**